engine/src/main/java/com/google/android/fhir/db/impl/DatabaseImpl.kt

```diff
@@ -128,6 +130,34 @@ internal class DatabaseImpl(context: Context, private val iParser: IParser, data
     localChangeDao.discardLocalChanges(token)
   }
 
+  override suspend fun syncUpload(
+    upload: suspend (List<SquashedLocalChange>) -> List<LocalChangeToken>
+  ) {
+    db.withTransaction {
+      var localChanges = getAllLocalChanges()
+      while (localChanges.isNotEmpty()) {
+        upload(localChanges).forEach { deleteUpdates(it) }
+      }
```

Comment:
Lets not call user code in a transaction, if they try to access back into the database, bad things might happen.

```suggestion
    var localChanges = getAllLocalChanges()
      while (localChanges.isNotEmpty()) {
         val uploaded = upload(localChanges)
         deleteUpdates(uploaded)
         localChanges = getAllLocalChanges()
      }
```

now that suggestion hits into a race condition hence we need to have a mutex around this to avoid concurrent access.

Reply:
done